accepted/000070-cluster_awareness.md

```diff
@@ -0,0 +1,309 @@
+
+In the cluster overview, when listing individual nodes of the cluster, Uyuni/SUSE Manager will also offer the possibility of registering an individual node to Uyuni/SUSE Manager (with System > Bootstrapping).
+
+It is common between all analyzed cluster products that rebooting and package modification/removal are two forbidden actions on all nodes of the cluster.
+
+By default, Uyuni/SUSE Manager will put the minion in system lock when one of the following product is detected upon registration:
+
+* CaaSP: /etc/products.d/caasp.prod + pattern `pattern-caasp-Node`
+* SES: /etc/products.d/ses.prod
+* SLE-HA/SHAP: /etc/products.d/SLES_SAP.prod, /etc/products.d/sle-ha.prod
+
+When a system is locked, no action will be executed. System lock may be disabled by the user by disabling the corresponding Salt Formula.
+
+It will be possible to define an allowed list of actions that Uyuni/SUSE Manager can issue on a specific node by implementing a combination of selective blackout (https://github.com/uyuni-project/uyuni-rfc/pull/31/) and package locking.
+
+A product can define a list of:
+- an allowed list of actions in Uyuni/SUSE Manager action terms
+- packages or patterns to lock at the minion level
+
+This information must be provided by the product and must be inserted into Uyuni/SUSE Manager database.
+Upon bootstrapping, Uyuni/SUSE Manager will issue the required package locks and the allowed list of actions will be not part of the blackout.
+Example:
+- CaaSP define the `zypper lock` the `patterns-caasp*` pattern
+- Uyuni/SUSE Manager allows package install, modify, removal actions
+
+#### Cluster node as special citizens in the minion domain
+
+All systems add-on types will be disabled for installation, except for the Monitoring add-on.
+When deleting a system, a cleanup of salt-minion must not be done (requested by SES).
+
+All other actions will be inhibited by the system lock and will require special handling by being implemented in the cluster provider manager level.
+
+#### System groups
+
+When defining a new cluster, Uyuni/SUSE Manager also creates a system group named `<type><name>` in which all nodes of the cluster will be inserted.
+
+## Integration requirements from an external product perspective
+
+From the plugin perspective, a product needs to define:
+- a new cluster type in Uyuni/SUSE Manager with a specific configuration (cluster type <-> cluster provider manager)
+- provide a system add-on type (or a Salt Formula) that provisions the management node with cluster management software
+
+When a new cluster product type is available in Uyuni/SUSE Manager, the association between the cluster type and its cluster provider manager is stored in the database.
+
+Summary of what a product must provide to Uyuni/SUSE Manager:
+
+* CaaSP:
+  - Salt Formula to install CaaSP software on the management node
+  - Salt Formula to install CaaSP software for control-plane/workers
+  - Cluster provider manager in a Salt module: skuba
```

Comment:
It is not clear to me what a cluster provider manager is at this point. My naive idea was that a CPM was a self-contained, Salt-based plugin in (eg. directory structure) containing modules, formulas and metadata that could be installed onto a SUSE Manager/Uyuni Server to "enable" it to manage a given cluster.

According to that vision, in this example, the two formulas above and the module would be part of one such CPM directory/package, and the SUSE Manager/Uyuni code base would not really be aware of product specifics, unless though the CPM itself.

I am not sure if this is the idea reflected in this RFC at this point, can you illustrate?

Reply:
Same question here. It would probably help to document somewhere which different components will be needed to cover all parts of the cluster management (specific repos, Salt states, tools to be installed for the cluster manager, icons, metadata, ...).

Reply:
I rephrased it into the RFC as follows:

Summary of what a product must provide to Uyuni/SUSE Manager:

* CaaSP: Salt based cluster provider manager ("`skuba`")
* SES: Salt based cluster provider manager ("`cephadm`")
* SLE-HA/SHAP: Salt based cluster provider ("`crm`")

It is intendend that the cluster provider manager package must at least contain:

- The Salt Formula to deploy the management node
- All Salt modules and states to implement the cluster actions described above